Gems/RecastNavigation/Code/Source/Misc/RecastNavigationDebugDraw.cpp

```diff
@@ -0,0 +1,116 @@
+    }
+
+    void RecastNavigationDebugDraw::vertex(float x, float y, float z, unsigned color,
+        [[maybe_unused]] const float u, [[maybe_unused]] const float v)
+    {
+        AddVertex(x, y, z, color);
+    }
+
+    void RecastNavigationDebugDraw::end()
+    {
+        constexpr AZ::Crc32 viewportId = AzFramework::g_defaultSceneEntityDebugDisplayId;
+        AzFramework::DebugDisplayRequestBus::BusPtr debugDisplayBus;
+        AzFramework::DebugDisplayRequestBus::Bind(debugDisplayBus, viewportId);
+        AzFramework::DebugDisplayRequests* debugDisplay = AzFramework::DebugDisplayRequestBus::FindFirstHandler(debugDisplayBus);
+        if (!debugDisplay)
+        {
+            return;
+        }
+
+        switch (m_currentPrim)
+        {
+        case DU_DRAW_POINTS:
+            for (auto& i : m_verticesToDraw)
+            {
+                AZ::Color color = AZ::Color::CreateZero();
+                color.FromU32(i.second);
+                debugDisplay->SetColor(color);
+
+                debugDisplay->DrawPoint(i.first, 1);
+            }
+            break;
+        case DU_DRAW_TRIS:
+            for (size_t i = 2; i < m_verticesToDraw.size(); i += 3)
+            {
+                AZ::Color color = AZ::Color::CreateZero();
+                color.FromU32(m_verticesToDraw[i].second);
+
+                debugDisplay->DrawTriangles(
+                    {
+                    m_verticesToDraw[i - 2].first,
+                    m_verticesToDraw[i - 1].first,
+                    m_verticesToDraw[i - 0].first
+                    }, color);
+            }
+            break;
+        case DU_DRAW_QUADS:
+            for (size_t i = 3; i < m_verticesToDraw.size(); i += 4)
+            {
+                AZ::Color color = AZ::Color::CreateZero();
+                color.FromU32(m_verticesToDraw[i].second);
+                debugDisplay->SetColor(color);
+
+                debugDisplay->DrawQuad(
+                    m_verticesToDraw[i - 3].first, m_verticesToDraw[i - 2].first,
+                    m_verticesToDraw[i - 1].first, m_verticesToDraw[i - 0].first);
+            }
+            break;
+        case DU_DRAW_LINES:
+            for (size_t i = 1; m_drawLines && i < m_verticesToDraw.size(); i++)
+            {
+                AZ::Color color0 = AZ::Color::CreateZero();
+                color0.FromU32(m_verticesToDraw[i].second);
+
+                AZ::Color color1 = AZ::Color::CreateZero();
+                color1.FromU32(m_verticesToDraw[i].second);
+
+                debugDisplay->DrawLine(m_verticesToDraw[i - 1].first, m_verticesToDraw[i].first,
+                    color0.GetAsVector4(), color1.GetAsVector4());
+            }
+            break;
+        }
+    }
+
+    void RecastNavigationDebugDraw::AddVertex(float x, float y, float z, unsigned color)
+    {
+        const float temp[3] = { x, y, z };
+        const RecastVector3 v(temp);
+        m_verticesToDraw.emplace_back(v.AsVector3(), color);
```

Comment:
comment: This bit of code is doing multiple swaps of z and y and might be easier to read if the naming was clearer (unless we do intentionally want to try and hide that as an implementation detail, it just feels all too easy that it'll leak out and cause confusion). Might not be a big deal, it's just something that is very hard to change later 😛